presto-main/src/main/java/com/facebook/presto/cost/HistogramCalculator.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.spi.statistics.ConnectorHistogram;
+import com.facebook.presto.spi.statistics.Estimate;
+import com.google.common.math.DoubleMath;
+
+import java.util.Optional;
+
+import static java.lang.Double.isFinite;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.min;
+
+public class HistogramCalculator
+{
+    private HistogramCalculator()
+    {}
+
+    /**
+     * Calculates the "filter factor" corresponding to the overlap between the statistic range
+     * and the histogram distribution.
+     * <br>
+     * The filter factor is a fractional value in [0.0, 1.0] that represents the proportion of
+     * tuples in the source column that would be included in the result of a filter where the valid
+     * values in the filter are represented by the {@code range} parameter of this function.
+     *
+     * @param range the intersecting range with the histogram
+     * @param histogram the source histogram
+     * @param totalDistinctValues the total number of distinct values in the domain of the histogram
+     * @param useHeuristics whether to return heuristic values based on constants and/or distinct
+     * value counts. If false, {@link Estimate#unknown()} will be returned in any case a
+     * heuristic would have been used
+     * @return an estimate, x, where 0.0 <= x <= 1.0.
+     */
+    public static Estimate calculateFilterFactor(StatisticRange range, ConnectorHistogram histogram, Estimate totalDistinctValues, boolean useHeuristics)
```

Comment:
Nit: Is it better to have this control given to the ConnectorHistogram itself?

Reply:
I think that would lead to a lot of duplicate code in the implementations. At the core, histograms are able to answer basic questions about the data distribution which are the `cumulativeProbability` and `inverseCumulativeProbability` methods on the `ConnectorHistogram` interface. As long as those two methods can be implemented properly by a histogram you can answer any question the optimizer needs for comparison cost calculations.

Is there a specific reason that you think this should be handled by the individual implementations? Pushing this down to the implementation could not only have more potential for duplicate code but also lead to more errors if the implementations aren't written correctly.

Reply:
Sounds good, I was wondering adding this  as a default method but as you said its the same for all histograms